apis/management.cattle.io/v3/pipeline_types.go

```diff
@@ -0,0 +1,224 @@
+	TriggerUserName string   `json:"triggerUserName,omitempty" norman:"type=reference[user]"`
+	Pipeline        Pipeline `json:"pipeline,omitempty" norman:"required"`
+}
+
+type PipelineExecutionStatus struct {
+	Commit  string        `json:"commit,omitempty"`
+	State   string        `json:"state,omitempty"`
+	Started string        `json:"started,omitempty"`
+	Ended   string        `json:"ended,omitempty"`
+	Stages  []StageStatus `json:"stages,omitempty"`
+}
+
+type StageStatus struct {
+	State   string       `json:"state,omitempty"`
+	Started string       `json:"started,omitempty"`
+	Ended   string       `json:"ended,omitempty"`
+	Steps   []StepStatus `json:"steps,omitempty"`
+}
+
+type StepStatus struct {
+	State   string `json:"state,omitempty"`
+	Started string `json:"started,omitempty"`
+	Ended   string `json:"ended,omitempty"`
+}
+
+type SourceCodeCredentialSpec struct {
+	ClusterName    string `json:"clusterName" norman:"required,type=reference[cluster]"`
+	SourceCodeType string `json:"sourceCodeType,omitempty" norman:"required,options=github"`
+	UserName       string `json:"userName" norman:"required,type=reference[user]"`
+	DisplayName    string `json:"displayName,omitempty" norman:"required"`
+	AvatarURL      string `json:"avatarUrl,omitempty"`
+	HTMLURL        string `json:"htmlUrl,omitempty"`
+	LoginName      string `json:"loginName,omitempty"`
+	AccessToken    string `json:"accessToken,omitempty"`
+}
+
+type SourceCodeCredentialStatus struct {
+}
+
+type SourceCodeRepositorySpec struct {
+	ClusterName              string   `json:"clusterName" norman:"required,type=reference[cluster]"`
+	SourceCodeType           string   `json:"sourceCodeType,omitempty" norman:"required,options=github"`
+	UserName                 string   `json:"userName" norman:"required,type=reference[user]"`
+	SourceCodeCredentialName string   `json:"sourceCodeCredentialName,omitempty" norman:"required,type=reference[sourceCodeCredential]"`
+	URL                      string   `json:"url,omitempty"`
+	Permissions              RepoPerm `json:"permissions,omitempty"`
+	Language                 string   `json:"language,omitempty"`
+}
+
+type SourceCodeRepositoryStatus struct {
+}
+
+type RepoPerm struct {
+	Pull  bool `json:"pull,omitempty"`
+	Push  bool `json:"push,omitempty"`
+	Admin bool `json:"admin,omitempty"`
+}
+
+type PipelineExecutionLogSpec struct {
+	ProjectName string `json:"projectName" yaml:"projectName" norman:"required,type=reference[project]"`
+
+	PipelineExecutionName string `json:"pipelineExecutionName,omitempty" norman:"type=reference[pipelineExecution]"`
+	Stage                 int    `json:"stage,omitempty" norman:"min=1"`
+	Step                  int    `json:"step,omitempty" norman:"min=1"`
+	Line                  int    `json:"line,omitempty"`
+	Message               string `json:"message,omitempty"`
+}
+
+type AuthAppInput struct {
+	Type         string `json:"type,omitempty" norman:"type=string,required"`
```

Comment:
I recall type having special meaning in norman API, and we should avoid this name as a field. We've had some problem using Type field in `ingress` embedded types, so it had to be fixed: https://github.com/rancher/types/pull/185/commits/c00024288496c10960dc0bcbbcd7581ac097fa56. Can you double check if it won't result in any problems in your case.

Reply:
The Input object can be parsed in actionHandler apart from the framework. But I think it's a good idea to avoid using the type field to keep consistent with other places. I've updated the commit.